### PR TITLE
fix(plugins): stop root-declared components copying the plugin into itself (closes #2556)

### LINKED ISSUE #2556: [BUG] `"skills": ["./"]` in plugin.json crashes install with "maximum recursion depth exceeded"
A `plugin.json` declaring `"skills": ["./"]` — the documented Claude Code way to say "this plugin's single skill is at the plugin root" — makes `apm install` fail with `maximum recursion depth exceeded`. The plugin never installs.

The declaration is legal on the Claude Code side. The [plugins reference](https://code.claude.com/docs/en/plugins-reference) documents it directly: when a skill path points at a directory holding `SKILL.md` — "for example `"skills": ["./"]` pointing to the plugin root" — the frontmatter `name` determines the skill's invocation name. So this is a shape APM will meet in the wild on any marketplace that ships single-skill plugins.

## Reproduction

Minimal, no network:

```bash
mkdir -p plug/.claude-plugin consumer
printf -- '---\nname: hello\ndescription: Minimal solo skill.\n---\n\n# hello\n' > plug/SKILL.md
printf '{"name":"p","version":"1.0.0","skills":["./"]}' > plug/.claude-plugin/plugin.json

cd consumer && apm install ../plug --target claude
```

Output:

```text
[>] Resolving ../plug...
[i] Targets: claude  (source: --target flag)
  [+] ../plug (local)
[i] Removed apm.yml created by the failed install.
[x] Failed to install APM dependencies: Failed to resolve APM dependencies:
maximum recursion depth exceeded
[i] Run with --verbose for detailed diagnostics
[x] Install failed after 1.9s.
```

`--verbose` adds no diagnostic beyond the same one-line message.

Also reproduces from a public marketplace over the network, via the `.claude-plugin/marketplace.json` intake path:

```bash
apm marketplace add DgxSparkLabs/marketplace-template
apm install skill-example-single@dgxsparklabs-template-marketplace --target claude
```

```text
  [x] 1 package failed:
    +- marketplace-template-skill-example-single -- Failed to download dependency
       DgxSparkLabs/marketplace-template: Subdirectory is not a valid APM package or
       Claude Skill: Failed to process Claude plugin: maximum recursion depth exceeded
```

That repository is mine and I am changing the layout that triggers it, so the local repro above is the durable one.

## Mechanism

`_map_plugin_artifacts` stages declared components into `.apm/`, which lives **inside** the plugin directory. When the declared entry resolves to the plugin root itself, source and destination-parent are the same directory, and `shutil.copytree` walks into the tree it is writing:

```text
plug/                              <- copy FROM here (because "./")
├── SKILL.md
├── .claude-plugin/
└── .apm/skills/
    └── plug/                      <- copy TO here, inside the source
        └── .apm/skills/plug/
            └── .apm/skills/plug/  <- until the stack is exhausted
```

`src/apm_cli/deps/plugin_parser.py` (main @ `3aa0365`), the list branch:

```python
is_custom_list = isinstance(manifest.get("skills"), list)
if is_custom_list and skill_dirs:
    target_skills.mkdir(parents=True, exist_ok=True)
    for d in skill_dirs:
        nested = target_skills / d.name
        if _is_same_path(d, nested):     # <- catches src == dst only
            continue
        shutil.copytree(d, nested, ignore=ignore_non_content, dirs_exist_ok=True)
```

`_is_same_path` compares resolved equality. Here `d` is the plugin root and `nested` is `<plugin root>/.apm/skills/<name>` — a **descendant**, not an equal — so the guard does not fire. The comment above the helper anticipates only the equal case ("Copying onto self raises `shutil.SameFileError`").

Frames, captured by calling the entry point directly:

```text
RecursionError; total frames: 1001
  src/apm_cli/deps/plugin_parser.py:267 in synthesize_apm_yml_from_plugin
  src/apm_cli/deps/plugin_parser.py:793 in _map_plugin_artifacts
  ... shutil.copytree recursing into its own destination
```

## Scope

Confirmed by differential testing — three plugins differing only in the declaration and file placement:

| `plugin.json` `skills` | Layout | Result |
| ---------------------- | ------ | ------ |
| `["./"]` | root `SKILL.md` | **`maximum recursion depth exceeded`** |
| *(key absent)* | root `SKILL.md` | Installs. Deploys under the **plugin directory name**, ignoring frontmatter `name` |
| `["./skills/"]` | `skills/<name>/SKILL.md` | Installs correctly as `<name>` |

So the trigger is the declared `"./"` entry specifically, not root-`SKILL.md` layouts in general.

Not affected, checked explicitly:

- The source plugin directory is not polluted — APM works on a copy, and `plug/` still contains exactly its two original directories afterwards.
- No oversized tree is left in `~/.apm`.
- Rollback is otherwise clean; only empty `apm_modules/` and `.claude/` directories remain in the consumer.

## Suggested fix

Make the guard a **containment** check rather than an equality check, so a destination nested inside its own source is never used as a `copytree` target. Two placements, whichever fits the design:

1. In the loop — when `nested` would land inside `d`, stage to a temporary directory outside the plugin root and move the result into place.
2. At classification — treat a declared entry that resolves to the plugin root as the "this entry *is* one skill" case and stage it under its resolved skill name from a copy taken before `.apm/` exists.

Either way, the name used should be the frontmatter `name`, since that is what Claude Code uses for a root-`SKILL.md` plugin and what a consumer will type.

## Relationship to open work

- **#2537** — declared-vs-discovered skill sets disagreeing. Same file, later stage: that bug is about which skills win, this one crashes before any set is produced.
- **#2530 / #2536** — `--skill` enumeration on subdirectory refs. From #2536's description, declared entries are classified so that "one carrying its own `SKILL.md` keeps its name" — which reads like the same branch that recurses here, so I do not expect #2536 to fix this. Worth a check by whoever owns that PR; if it does fix it, this can close as a duplicate with a regression test.
- **#760** (`harden(plugin): enforce plugin-root containment for manifest paths`) — established that declared paths must stay inside the plugin root. `"./"` satisfies that check and then breaks on the copy, so the containment work has a gap on the write side.

A regression test asserting `"skills": ["./"]` installs would cover all three.

## Environment

- `apm` built from `main` at `3aa0365`, reported as `0.28.0 (3aa0365)`; also reproduced with `apm-cli` 0.28.0 from PyPI.
- Windows 11, Python 3.14.
- Reproduced with both a local path dependency and a remote `.claude-plugin/marketplace.json` marketplace.

*Disclosure: drafted with AI assistance; every command shown was executed against the versions named above and its output captured verbatim.*

### PR DESCRIPTION
**Pull request by YoraiLevi**
Closes #2556.

## Root cause

`"skills": ["./"]` is the documented Claude Code single-skill shape, so APM meets it on any marketplace shipping solo plugins. The staging root `.apm/` lives inside the plugin and every destination in `_map_plugin_artifacts` is under it, so a component declared at the plugin root makes the source an ancestor of its own destination. The copy then walks into the tree it is writing:

```text
plug/                              <- copy FROM here (because "./")
├── SKILL.md
├── .claude-plugin/
└── .apm/skills/
    └── plug/                      <- copy TO here, inside the source
        └── .apm/skills/plug/ ...
```

`_is_same_path` only catches `src == dst`. Here `dst` is a **descendant** of `src`, so the guard never fires. The comment above the helper anticipates only the equal case.

## The issue under-reported the scope

The report covers `skills`. Probing all four component types against clean `main` @ `3aa0365`:

| Component | Behaviour before this PR |
| --------- | ------------------------ |
| `skills` | `RecursionError` — install fails |
| `agents` | `RecursionError` — install fails |
| `hooks` | `RecursionError` — install fails |
| `commands` | **No exception.** Writes without bound — 3301 files and still growing after four minutes, from a two-file plugin |

`commands` is the worse one and the reason the fix is not skills-only: it walks with `rglob` rather than `copytree`, and `rglob` is lazy, so it keeps yielding the files the loop has just written.

## Fix

Exclude the staging root from component copies.

That is sufficient **and** complete: sources are constrained to the plugin root by `_is_within_plugin`, destinations are always under `apm_dir`, so `.apm/` is the only path from a permitted source back down to its own destination. Blocking that one edge makes every case terminate — no temp-dir staging, no move-into-place, no per-component special cases.

- **`ignore_non_content_and_staging`** composes the existing callback and drops the staging directory when copying from its parent. Derived from `apm_dir` rather than hardcoding `.apm`, so it holds if the staging root ever moves.
- **`_under_staging`** gives the `commands` walk an explicit skip, since an ignore-callback cannot reach `rglob`.

No behaviour change for any manifest that does not declare a component at the plugin root: the callback only drops an entry when the directory being copied *is* the staging root's parent.

## Interaction with #2536

#2536 rewrites this function, so worth flagging: from its description, declared entries are classified so that "one carrying its own `SKILL.md` keeps its name" — which is the branch that recurses here, so I do not think #2536 fixes this on its own. If it turns out to, the tests in this PR are still worth keeping as a regression guard. Happy to rebase on top of it in whichever order you prefer.

Also related: #760 established that declared paths must stay inside the plugin root. `"./"` satisfies that check and then breaks on the write, so this is the containment gap on the write side.

## Not addressed here

Two things a root-declared component still does. Both are pre-existing and both feel like they belong with #2537's "which skills win" question rather than with a crash fix — say the word and I will fold either in:

- The staged skill takes its name from the plugin directory (`d.name`), not the frontmatter `name` Claude Code invokes it by.
- Packaging files (`.claude-plugin/`, `apm.yml`) ride along into the staged component, because at the plugin root everything is "content".

## Verification

**New tests** — `TestRootDeclaredComponents`, 7 cases, parameterized across all four component types. Confirmed failing on clean `main`: the six non-hanging cases fail with `RecursionError`; the `commands` case hangs rather than failing, which is called out in the class docstring so nobody is surprised by it later.

```console
$ pytest tests/unit/test_plugin_parser.py -k RootDeclared -q     # on this branch
7 passed

$ pytest tests/unit/test_plugin_parser.py -k "RootDeclared and not commands" -q   # on clean main
6 failed
```

**End to end** — the minimal repro from the issue:

```console
$ apm install ../plug --target claude
  |-- Skill integrated -> .claude/skills/
[*] Installed 1 APM dependency in 0.3s.
```

**No new failures** — `tests/unit/deps tests/unit/install test_plugin_parser test_symlink_containment test_ignore_non_content test_plugin_exporter_schema`: 3347 passed. This Windows box has 42 pre-existing failures in that set; on the first commit the failure list was byte-identical to clean `main` run with the same command, so the diff was zero.

Re-running after the review commit gave 41 — one *fewer*. The difference is `test_direct_dep_failure.py::TestDirectDepFailLoud::test_pipeline_raises_direct_dependency_error`, which passes 3/3 in isolation and is unrelated to anything here; it looks order-dependent or flaky in the full-set run rather than fixed by this PR. Calling it out so the count is not read as an improvement.

**`TestIgnoreNonContentSourceGuard` still passes** — the composing helper keeps each `copytree` site's `ignore_non_content` reference, which is what that AST guard counts.

Environment: Windows 11, Python 3.14, branched from `3aa0365`. I did not run the full suite (it exceeds ten minutes here) or the integration suite — worth a CI pass before merge.

*Disclosure: drafted with AI assistance; every command shown was executed against the versions named above and its output captured verbatim.*

### CODE REVIEW

**Comment by YoraiLevi** [comment]
@microsoft-github-policy-service agree company="Yorai Levi"

File: tests/unit/test_plugin_parser.py [hunk 1]
```diff
@@ -1411,3 +1411,90 @@ def test_malformed_apm_yml_fallback_surfaces_warning(self, tmp_path):
...
+            plugin_dir, apm_dir, manifest={"name": "p", component: ["./"]}
+        )
+
+        # Bounded output is the assertion: the source holds two files, so a
+        # terminating copy cannot produce an order of magnitude more.
+        produced = [p for p in apm_dir.rglob("*") if p.is_file()]
+        assert len(produced) <= 8, (
+            f"{component}: {len(produced)} files staged from a 2-file plugin — "
+            f"the copy is consuming its own output"
+        )
```
[the remark this answers is not in the document]
Comment (YoraiLevi) [reply tests/unit/test_plugin_parser.py:1628]: Done in 24bde6f — the em dash was the only non-ASCII character in the diff, and it is now `--`.

Flagging the stated reason though, so it does not get cited as precedent: there is no repo-wide printable-ASCII rule for source files. `main` @ `3aa0365` carries 8,599 non-ASCII characters across 226 files under `src/` and `tests/`, and `scripts/check_diagnostic_ascii_owner.py` is a narrow architecture boundary requiring two specific diagnostic call sites to delegate to `printable_ascii_text` when rendering untrusted package and agent names — not a lint on source text. `tests/unit/test_plugin_parser.py` already contained two em dashes before this PR.

The change is still right on the merits: `plugin_parser.py` uses ASCII `--` in all nine of its prose dashes and no em dashes, and this string is a pytest assertion message that prints on failure, where a cp1252 console is a real hazard.

File: src/apm_cli/deps/plugin_parser.py [hunk 2]
```diff
@@ -700,6 +700,41 @@ def _map_plugin_artifacts(
...
+            if at_staging_parent:
+                dropped.append(_staging_name)
+        return dropped
+
+    def _under_staging(p: Path) -> bool:
+        """True when *p* lies inside the staging root."""
+        try:
+            return p.resolve().is_relative_to(apm_dir.resolve())
+        except OSError:
+            return False
```
[the remark this answers is not in the document]
Comment (YoraiLevi) [reply src/apm_cli/deps/plugin_parser.py]: Good catch — fixed in 24bde6f. `apm_dir.resolve()` is now hoisted to `_staging_root` alongside `_staging_parent` and `_staging_name`, so the walk resolves it once instead of once per yielded path.

`p.resolve()` per candidate stays, since each path still has to be resolved to compare it.

**Comment by YoraiLevi** [comment]
@microsoft-github-policy-service agree

**Comment by sergio-sisternes-epam** [comment]
## APM Review Panel: `ship_with_followups`

> Fixes infinite recursion and unbounded writes when installing root-declared plugins (skills: [./]), unblocking the most common Claude Code single-skill shape.

cc @YoraiLevi @danielmeppiel -- a fresh advisory pass is ready for your review.

All nine panelists converge on ship. The fix is minimal, architecturally sound (python-architect), net-positive for security (supply-chain), and unblocks a critical adoption path (growth-hacker). No panelist raised a blocking finding.

The test-coverage-expert flagged two missing unit-test surfaces: `_bounded_read_json` (a security boundary defending against JSON bombs) and `_validate_declared_component_paths` (an error-message contract). Both carry `outcome: missing` evidence on security-relevant code paths. Per arbitration weighting rules, missing tests on a secure-by-default surface rank above opinion-only recommended findings -- these should be the first follow-up items. However, because the primary regression path (the self-copy loop) IS tested with 7 cases, the absence of these ancillary tests does not block ship.

The doc-writer and growth-hacker both independently flagged the missing CHANGELOG entry. This is a valid gap: the fix resolves a silent-churn event for Claude Code users and the next release note must name it explicitly. The PR author or maintainer should add the entry before merge.

**Aligned with:** Portable by manifest -- `skills: [./]` is the canonical single-skill manifest shape; fix ensures it works as documented. Secure by default -- adds bounded JSON reader and staging-root exclusion, reducing DoS and unbounded-write attack surface. Pragmatic as npm -- `apm install` now handles the most common Claude Code plugin shape without user workarounds. OSS community driven -- external contributor PR fixing a real user-reported bug (#2556); fast triage reinforces community trust.

**Growth signal.** Claude Code interop is a top-of-funnel vector. This bug was silently killing conversions for the most common single-skill plugin shape. CHANGELOG entry should explicitly name "Claude Code single-skill shape" to create a searchable breadcrumb for users who already churned.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---|---|---|---|
| Python Architect | 0 | 0 | 1 | Clean, minimal fix composing existing ignore callback with staging-root exclusion; architecture is sound. |
| CLI Logging Expert | -- | -- | -- | Inactive: No CLI output changes in this PR. No logging concerns. |
| DevX UX Expert | 0 | 0 | 0 | No CLI surface, help text, or first-run flow changes. The new error messages are well-crafted. No UX concerns. |
| Supply Chain Security | 0 | 0 | 1 | Net security improvement. Adds bounded JSON reader, validates declared component paths, excludes staging root. No new attack surface. |
| OSS Growth Hacker | 0 | 1 | 1 | This fix unblocks a critical adoption path -- Claude Code's single-skill shape is a high-traffic entry point. Ship it. |
| Auth Expert | -- | -- | -- | Inactive: PR touches only plugin_parser.py path-copy logic; no auth surfaces affected. |
| Doc Writer | 0 | 1 | 1 | The PR fixes a user-observable bug but ships no CHANGELOG entry and no documentation update. |
| Test Coverage Expert | 0 | 1 | 1 | Core regression fix has strong unit tests (7 cases). Two new defensive code paths lack tests. |
| Performance Expert | 0 | 0 | 1 | No meaningful performance regression. resolve() calls are guarded and bounded. |

> B = blocking-severity findings, R = recommended, N = nits.
> Counts are signal strength, not gates. The maintainer ships.

### Top 5 follow-ups

1. **[Test Coverage Expert]** Add unit tests for `_bounded_read_json` oversize/malformed rejection -- Security boundary (DoS defense) with outcome:missing evidence on a secure-by-default surface. Highest priority per arbitration weighting.
2. **[Test Coverage Expert]** Add parametrized tests for `_validate_declared_component_paths` error paths -- Error-message contract with outcome:missing evidence; locks actionable DX for bad manifests.
3. **[Doc Writer]** Add CHANGELOG [Unreleased] entry naming the fix and Claude Code shape -- Every observable behavior change needs a CHANGELOG line; also serves growth signal for Claude Code community.
4. **[Supply Chain Security]** Document `_under_staging` OSError fallback direction -- One-line comment clarifies why False is safe given the preceding `is_symlink()` guard.
5. **[Python Architect]** Extract ignore-factory to a named helper for reuse and testability -- Minor refactor; improves standalone testability if future integrators need the same guard.

### Architecture

```mermaid
classDiagram
    direction LR
    class PluginParser {
      +parse_plugin()
    }
    class _map_plugin_artifacts {
      <<Function>>
      +apm_dir Path
      +plugin_path Path
      +manifest dict
    }
    class ignore_non_content {
      <<Callback>>
      +__call__(dir, contents) list~str~
    }
    class ignore_non_content_and_staging {
      <<Callback - Composed>>
      +__call__(dir, contents) list~str~
    }
    class _under_staging {
      <<Predicate>>
      +__call__(p Path) bool
    }
    class _is_within_plugin {
      <<SecurityGuard>>
      +__call__(path) bool
    }
    class shutil_copytree {
      <<stdlib>>
    }
    PluginParser ..> _map_plugin_artifacts : calls
    _map_plugin_artifacts *-- ignore_non_content_and_staging : creates
    _map_plugin_artifacts *-- _under_staging : creates
    ignore_non_content_and_staging ..> ignore_non_content : delegates
    _map_plugin_artifacts ..> shutil_copytree : invokes
    _map_plugin_artifacts ..> _is_within_plugin : guards paths
    shutil_copytree ..> ignore_non_content_and_staging : ignore param
    class ignore_non_content_and_staging:::touched
    class _under_staging:::touched
    classDef touched fill:#fff3b0,stroke:#d47600
```

```mermaid
flowchart TD
    A["apm install plugin"] --> B["_map_plugin_artifacts()\nplugin_parser.py"]
    B --> C["Resolve _staging_root = apm_dir.resolve()"]
    C --> D{"Component type?"}
    D -->|skills/agents/hooks| E["shutil.copytree(src, dst,\nignore=ignore_non_content_and_staging)"]
    D -->|commands| F["rglob walk over source"]
    E --> G{"ignore callback: dir == _staging_parent\nAND .apm in contents?"}
    G -->|Yes| H["Skip .apm/ subtree"]
    G -->|No| I["Copy file normally"]
    F --> J{"_under_staging(file)?"}
    J -->|Yes| K["Skip file"]
    J -->|No| L["_copy_command_file()"]
    style E fill:#fff3b0,stroke:#d47600
    style F fill:#fff3b0,stroke:#d47600
```

### Recommendation

Ship now with the CHANGELOG entry added before merge (trivial ask for the author). The two missing-test follow-ups should land as a fast-follow PR but do not block this fix -- the primary regression path is well-tested and the bug is actively blocking Claude Code adopters. No security, architecture, or UX concerns.

---

<details>
<summary>Full per-persona findings</summary>

#### Python Architect

- **[nit]** Consider extracting ignore-factory to a named helper for testability at `src/apm_cli/deps/plugin_parser.py:720`
  The closure `ignore_non_content_and_staging` captures `_staging_parent` and `_staging_name` from enclosing scope. A standalone factory `make_staging_aware_ignore(apm_dir, base_ignore)` would be independently unit-testable and reusable if future integrators need the same guard.

#### CLI Logging Expert -- inactive

Inactive: No CLI output changes in this PR. No logging concerns.

#### DevX UX Expert

No findings.

#### Supply Chain Security Expert

- **[nit]** `_under_staging` OSError fallback is safe-for-termination but not safe-for-exclusion at `src/apm_cli/deps/plugin_parser.py:735`
  Returning False on OSError means the path is NOT excluded. If rglob yields a dangling symlink path that fails to resolve, it would be copied. The preceding `is_symlink()` check guards this in practice but the fallback direction should be documented.
  *Suggested:* Add comment: `# False = do not skip; safe because the preceding is_symlink() guard already excludes unresolvable symlinks.`

#### OSS Growth Hacker

- **[recommended]** Add a story-shaped CHANGELOG entry that names the Claude Code ecosystem
  Claude Code users hitting RecursionError on `apm install` is a silent churn event. A CHANGELOG entry naming `skills: ["./"]` (common in Claude Code plugins) signals to the Claude Code community that APM actively supports their workflow.
  *Suggested:* Ensure CHANGELOG entry names "Claude Code single-skill shape" explicitly, not just the technical root cause.

- **[nit]** Consider a one-line mention in the troubleshooting or FAQ docs
  Users who already gave up will Google the traceback. A docs mention creates a searchable breadcrumb.

#### Auth Expert -- inactive

Inactive: PR touches only plugin_parser.py path-copy logic and its unit tests; no auth surfaces affected.

#### Doc Writer

- **[recommended]** CHANGELOG [Unreleased] section has no entry for this bug fix
  Every observable behavior change should have a CHANGELOG entry. This PR fixes a silent install failure for `skills: ["./"]`, which is the "documented Claude Code single-skill shape". Omitting it leaves the next release note incomplete.
  *Suggested:* Add under `## [Unreleased]`: `- Root-declared components (skills: ["./"], agents: ["./"] ...) no longer cause infinite recursion or unbounded writes during apm install. (closes #2556)`

- **[nit]** `reference/cli/install.md` does not mention `./` as a valid root-component path
  A reader authoring a single-skill Claude Code plugin who writes `skills: ["./"]` would not know from this page whether that form is valid.
  *Suggested:* Append: `A root component path ("./") is valid and deploys all content at the plugin root; the staging directory (.apm/) is excluded automatically.`

#### Test Coverage Expert

- **[recommended]** `_bounded_read_json` security boundary has no test for oversize or malformed JSON rejection
  grep found no matches for `_bounded_read_json` or `_MAX_PLUGIN_JSON` in tests/. This function caps plugin JSON at 5MB and converts RecursionError/MemoryError into ValueError. A unit test would lock this security boundary.
  *Proof (missing):* `tests/unit/test_plugin_parser.py::test_bounded_read_json_rejects_oversize` -- proves: Oversize plugin JSON is rejected before parse, preventing DoS

- **[nit]** `_validate_declared_component_paths` error paths have no direct test
  grep found no matches for `_validate_declared_component` or `DeclaredPluginComponentError` in tests/. The three distinct error raises are not directly exercised.
  *Proof (missing):* `tests/unit/test_plugin_parser.py::test_validate_declared_component_paths_empty_raises` -- proves: Empty component declarations produce actionable error messages

#### Performance Expert

- **[nit]** `_under_staging` does a full `resolve()` per file; could short-circuit with a str prefix check first
  Each `p.resolve()` is a `realpath(3)` syscall. For typical plugins N < 50 so cost is sub-millisecond. A cheap `str(p).startswith(str(_staging_root))` pre-check would skip the resolve for non-matching paths entirely.
  *Suggested:* Optional fast-path: `if not str(p).startswith(str(_staging_root)): return False` before `p.resolve().is_relative_to(...)`

</details>

<sub>This panel is advisory. It does not block merge. Re-apply the
`panel-review` label after addressing feedback to re-run.</sub>

**Comment by danielmeppiel** [comment]
## APM Review Panel: `needs_rework`

> Root-declared plugin components preserve portable manifests, but the staging guard omitted pre-positioned `.apm` commands and needed bounded lifecycle proof.

cc @YoraiLevi @sergio-sisternes-epam -- a fresh advisory pass is ready for your review.

The panel supports the narrow repair: mapping authority remains local to `_map_plugin_artifacts`, the four-component staging exclusion remains intact, and this stack incorporates #2536 without deciding #2537 precedence. The demonstrated pre-positioned command regression and lifecycle guardrail were folded into the current head.

**Aligned with:** portable-by-manifest root declarations retain their frontmatter content; staging exclusions preserve valid declared sources; install, reinstall, and uninstall leave a bounded deployed tree.

**Growth signal.** The included changelog entry makes the compatibility restoration discoverable without expanding authoring semantics.

### Panel summary

| Persona | B | R | N | Takeaway |
|---|---|---|---|---|
| Python Architect | 0 | 0 | 1 | Keep the composed staging guard local to artifact mapping. |
| CLI Logging | 0 | 0 | 0 | No CLI output surface changed. |
| DevX UX | 0 | 1 | 0 | Cover the install lifecycle at the user boundary. |
| Supply Chain Security | 1 | 0 | 0 | Preserve explicit pre-positioned command sources. |
| OSS Growth | 0 | 1 | 0 | Record the compatibility repair concisely. |
| Test Coverage | 0 | 1 | 0 | Add the bounded root-skill lifecycle regression. |
| Performance | 0 | 1 | 0 | Prune generated command output during traversal. |

> B = blocking-severity findings, R = recommended, N = nits. Counts are signal strength, not gates. The maintainer ships.

### Top 3 follow-ups

1. **[Supply Chain Security]** Restrict the generated-output exclusion so `.apm/custom-commands` remains a valid input. This was folded with a regression test.
2. **[Test Coverage]** Add a local root-skill install/reinstall/uninstall lifecycle proof. This was folded without asserting #2537 precedence.
3. **[OSS Growth]** Add a concise changelog entry. This is present in the stacked diff.

### Recommendation

The in-scope panel follow-ups are folded. Reassess the updated head with CI results; keep #2537 precedence out of this repair.

---

<details>
<summary>Full per-persona findings</summary>

#### Python Architect
- **[nit]** Keep the composed staging guard local to `_map_plugin_artifacts`.

#### CLI Logging Expert
No findings.

#### DevX UX Expert
- **[recommended]** The lifecycle guard now exercises the user-facing local install path.

#### Supply Chain Security Expert
- **[blocking]** Pre-positioned command sources were skipped by the first guard; the current head scopes pruning to generated output and covers this regression.

#### OSS Growth Hacker
- **[recommended]** Changelog coverage is present.

#### Auth Expert -- inactive
No authentication surface changed.

#### Doc Writer -- inactive
No documentation drift beyond the changelog repair.

#### Test Coverage Expert
- **[recommended]** The current head adds the requested local install/reinstall/uninstall lifecycle scenario.

#### Performance Expert
- **[recommended]** The command walk now prunes generated output rather than scanning and rejecting it file-by-file.
</details>

<sub>This panel is advisory. It does not block merge. Re-apply the `panel-review` label after addressing feedback to re-run.</sub>

**Comment by danielmeppiel** [comment]
## APM Spec Guardian: `fold_and_ship`

> Scope: **editorial-patch**; diff = +12/-4 lines across 1 file. Shocked-meter avg: **8.5/10**.

All four panelists report that the staging-boundary clarification is closed and identify no new findings. The advisory recommendation is to ship the editorial patch.

### Convergence

| Panel | Position | Shocked | New B | New R | New N |
|---|---|---:|---:|---:|---:|
| Spec Swagger Editor | ship | 8/10 | 0 | 0 | 0 |
| OCI Distribution Editor | ship | 8/10 | 0 | 0 | 0 |
| Package Manager Editor | ship | 9/10 | 0 | 0 | 0 |
| TAG Architect | ship | 9/10 | 0 | 0 | 0 |

> B = new blocking findings, R = new recommended, N = new nits.
> Counts are signal strength, not gates. The maintainer ships.

### Linter notes (1 check failed)

- **[11]** Product Python files are part of this PR; the canonical Python lint chain ran clean alongside the spec checks.

> **Linter handoff:** Verify the req-pr-007 canonical-source four-way bind and normative-statement count reconciliation across the required count sites.

---

<details>
<summary>Full per-panel findings</summary>

#### Spec Swagger Editor -- shocked_meter 8/10, confidence high

*Summary:* F1 is closed with a testable traversal-containment rule.

#### OCI Distribution Editor -- shocked_meter 8/10, confidence high

*Summary:* The clarified canonical-root and pre-traversal pruning contract closes the self-copy path.

#### Package Manager Editor -- shocked_meter 9/10, confidence high

*Summary:* F1 is closed without changing registry or lockfile contracts.

#### TAG Architect -- shocked_meter 9/10, confidence high

*Summary:* The requirement now supplies an implementable containment boundary.

</details>

<sub>This panel is advisory. Re-apply the `spec-review` label after addressing feedback to re-run.</sub>
